woop: uses combobox to navigate countries

Rolilink opened the pull request:
#### Changes proposed in this Pull Request
* Updating Country/Region dropdown to use ComboboxControl

<img width="1058" alt="Screen Shot 2022-01-10 at 2 25 40 PM" src="https://user-images.githubusercontent.com/1689238/148826683-807fded8-39a6-4704-9e59-eb3726fd37bb.png">

#### Testing instructions
* Apply D72720-code and sandbox the API.
* Navigate to `http://calypso.localhost:3000/start/woocommerce-install/store-address?site=<siteId>`
* Select a country (and region, if available) in the country dropdown
* Fill the rest and click continue
* With D72639-code it should send an http request to settings with `woocommerce_default_country` being the code of the selected country.

Related to #
* https://github.com/Automattic/wp-calypso/issues/59833

Rolilink committed:
woop: uses combobox to navigate countries

Rolilink committed:
woop: fixes search issues caused by repeated options

lsl commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -52,11 +61,12 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
-					<FormCountrySelect
+					<ComboboxControl
 						name="country_code"
+						label={ __( 'Country/Region', 'woocommerce-admin' ) }
```
```suggestion
						label={ __( 'Country / Region', 'woocommerce-admin' ) }
```

allilevine committed:
Add spacing to Country / Region label

Co-authored-by: Louis Laugesen <louis.laugesen@gmail.com>

allilevine committed:
Use new woocommerce/countries/regions endpoint

sirbrillig commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -52,11 +53,12 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
 
-					<FormCountrySelect
+					<ComboboxControl
 						name="country_code"
```
The `name` prop does not exist on `ComboboxControl`.

lsl replied:
d0ffd5a

sirbrillig commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -30,9 +28,12 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
 
-	const handleCountryChange: ChangeEventHandler< HTMLInputElement > = ( event ) => {
-		update( WOOCOMMERCE_DEFAULT_COUNTRY, event.target.value );
+	const handleCountryChange = ( value: string ) => {
```
The `onChange` callback for this function can sometimes be called with a `null` value, so you'll need to handle that here. Perhaps just with an early return?

lsl replied:
Null looks like they happen when you click the x reset button, just set this to empty string (c601fea), we'll be validating these values shortly.

sirbrillig commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -52,11 +53,12 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
 					/>
 
-					<FormCountrySelect
+					<ComboboxControl
```
This component does not yet have types in the package definition for `@wordpress/components` which is causing your type errors. I've added them in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58106. Once that merges and the calypso types are updated, this should pass the unit tests. The Calypso team can cause it to pass sooner if you ask!

lsl replied:
Thanks, we might do that. cc @p-jackson since we're midchat about this :laughing:

lsl committed:
No name field on combobox control props @sirbrillig

lsl committed:
Remove i18n contexts blocking translation reuse

lsl committed:
Handle null values (box reset)

lsl committed:
Width fix for combo box

lsl commented on client/signup/steps/woocommerce-install/step-store-address/style.scss:
```diff
@@ -2,3 +2,7 @@
 
+.components-combobox-control__suggestions-container {
+	// Remove `width: 100%;` https://github.com/WordPress/gutenberg/blob/0135921bcb72ceec629fa8f9e755891afab747b9/packages/components/src/combobox-control/style.scss#L32
+	width: inherit;
```
```suggestion
	width: unset;
```

lsl committed:
Unset probably safer

lsl commented:
Some kind of encoding happening on the output:

![Screenshot 2022-01-11 at 13-52-25 Add WooCommerce to your site — WordPress com](https://user-images.githubusercontent.com/811776/148873039-dbc4da98-5391-4ce2-8693-09cf11cff6ee.png)

sirbrillig commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -60,33 +60,33 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
-						name="country_code"
+					<ComboboxControl
+						label={ __( 'Country / Region' ) }
 						value={ get( WOOCOMMERCE_DEFAULT_COUNTRY ) }
```
I'm not sure why exactly, but TS believes `get( WOOCOMMERCE_DEFAULT_COUNTRY )` can return a `number`, which causes a type error since the `value` prop must be a string. Do you believe this should be a string? If so, can you cast it to a String here?

lsl replied:
Is it stemming from [this](https://github.com/Automattic/wp-calypso/pull/59505/files#diff-24648f1a5849a32e3be548a0a3991dc3cafc59a3574d553d9dd75de8225aed29R40)? `as string` should be fine here right? I don't think there is any cases where a non string can enter this setting.

lsl replied:
@retrofox is `blog_public` required in this hook? Can't see where/how it's used (still catching up after the break) - might be faster to just remove the def.

allilevine replied:
@sirbrillig What's the best way to catch errors like these going forward? I wasn't seeing a type error there. 🤔 

I cast it to a string here: https://github.com/Automattic/wp-calypso/pull/59876/commits/ab4b8ac7918a2eb27eb5d2a15405b87bd56c1749

> Can't see where/how it's used (still catching up after the break) - might be faster to just remove the def.

@lsl I'm not seeing this either but it seems like `get` should be able to return different types, so I'm hesitant to limit it to strings.

sirbrillig replied:
Good question! Sorry I didn't explain that earlier. You'd see the type errors if the types for `ComboboxControl` were available, but we're going to have to wait for the changes I merged there to be released and imported into calypso (and then rebased into this branch). Here's an example of when this was done recently: https://github.com/Automattic/wp-calypso/pull/59779

The reason I see them is that I manually added the types to my `node_modules` folder so I could check that they worked correctly, and that's why I was mentioning the errors to you here. This sort of thing is really weird and probably won't happen often.

sirbrillig replied:
> but we're going to have to wait for the changes I merged there to be released and imported into calypso (and then rebased into this branch)

Since it's not clear to me how long this process takes, you may want to ping the calypso team and ask them to mark these unit tests as passing (assuming the missing `ComboboxControl` is still the only new error). I'd do it but I don't have that power. 😅

allilevine replied:
> The reason I see them is that I manually added the types to my node_modules folder so I could check that they worked correctly, and that's why I was mentioning the errors to you here. 

Ahhh! Got it, thanks!

allilevine committed:
Fix rebase

allilevine committed:
Cast get( WOOCOMMERCE_DEFAULT_COUNTRY ) as string

obenland committed:
Remove reset button

That button seems defunc'd and also causes the input to be a different height than the rest. Clicking outside of the input field will reset it.

obenland committed:
Limit suggestion list width

Without this style the suggestions list will expand the width of the entire form. This will keep it at the width of the other input fields.

obenland reviewed (approved):
Looking good to me. 

Should we wait until https://github.com/Automattic/jetpack/pull/22248 made it to dotcom before merging it?

obenland commented:
This should be good to go now. 

@Automattic/team-calypso could we ask you to mark the unit tests as passing, [as suggested here](https://github.com/Automattic/wp-calypso/pull/59876#discussion_r782342538)?

tyxla commented:
> 

This was pending on #60002 essentially, but I've marked it as green now so you're unblocked to merge. Once #60002 lands, the related TS error that is being introduced here should get resolved.
